### PR TITLE
Use element comparison instead of path to verify source

### LINKED ISSUE #68: Use `CtElement` instead of `CtPath` strings for checking source file of an element
Reference: https://github.com/SpoonLabs/diffmin/pull/66#discussion_r644823057

Comparing elements will help in checking for updates to access modifiers as checking paths won't be sufficient there (there would be no change in path of the element as they are virtual elements).

### PR DESCRIPTION
**Pull request by algomaster99**
Fixes #68

### CODE REVIEW

File: src/test/java/com/diffmin/AppTest.java [hunk 1]
```diff
@@ -194,17 +196,16 @@ void should_apply_mix_operation_patches(TestResources sources) throws Exception
...
+            this.newRevisions = newRevisions;
         }
 
         @Override
         public void scan(CtElement element) {
-            if (!skipAssertionCheck(element, newRevisionPathStrings)) {
-                String elementPathString = element.getPath().toString();
-                if (newRevisionPathStrings.contains(elementPathString)
+            if (!skipAssertionCheck(element, newRevisions)) {
+                if (newRevisions.stream().anyMatch(modifiedElement -> modifiedElement == element)
```
Comment (algomaster99) [inline src/test/java/com/diffmin/AppTest.java:208]: I was expecting `contains` to have an option to check via `==` but couldn't find it. Had to resort to `anyMatch`.

Comment (slarse) [reply src/test/java/com/diffmin/AppTest.java:208]: You can use a set collection based around identity.

```java
    static class ElementSourceFileChecker extends CtScanner {
        private final Set<CtElement> newRevisions;

        ElementSourceFileChecker(List<CtElement> newRevisions) {
            this.newRevisions = Collections.newSetFromMap(new IdentityHashMap<>());
            this.newRevisions.addAll(newRevisions);
        }

[...]
```

```suggestion
                if (newRevisions.contains(element)
```

File: src/test/java/com/diffmin/AppTest.java [hunk 2]
```diff
@@ -236,23 +237,17 @@ private static boolean doesElementBelongToModifiedSet(CtElement element) {
...
 
         private static boolean isChildOfInsertedPath(
-                String elementPathString, Set<String> newRevisionPathStrings) {
-            return newRevisionPathStrings.stream()
-                    .anyMatch(
-                            modifiedPathString ->
-                                    !(modifiedPathString.equals(elementPathString))
-                                            && elementPathString.startsWith(modifiedPathString));
+                CtElement element, List<CtElement> newRevisions) {
+            return newRevisions.stream().anyMatch(element::hasParent);
```
Comment (slarse) [inline src/test/java/com/diffmin/AppTest.java:241]: This method shouldn't be static as it's using a field (`newRevisions`).

File: src/test/java/com/diffmin/AppTest.java [hunk 3]
```diff
@@ -236,23 +237,17 @@ private static boolean doesElementBelongToModifiedSet(CtElement element) {
...
-                CtElement element, Set<String> newRevisionPathStrings) {
+        private static boolean skipAssertionCheck(CtElement element, List<CtElement> newRevisions) {
             if (element == null
                     || element.isImplicit()
                     || !element.getPosition().isValidPosition()) {
                 return true;
             }
-            String elementPathString = element.getPath().toString();
-            return isChildOfInsertedPath(elementPathString, newRevisionPathStrings);
+            return isChildOfInsertedPath(element, newRevisions);
```
Comment (slarse) [inline src/test/java/com/diffmin/AppTest.java:250]: Idem; also shouldn't be static.

**Review by slarse — CHANGES_REQUESTED** [review]
Looks like it works as intended to me, I've some minor suggested changes if you want to make some improvements.

**Review by slarse — APPROVED** [review]
Very nice, LGTM!

**Pull request merged**